flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala

```diff
@@ -630,6 +628,24 @@ class FlinkChangelogModeInferenceProgram extends FlinkOptimizeProgram[StreamOpti
         }
 
       case normalize: StreamPhysicalChangelogNormalize =>
+        val tableIdentifier = normalize.tableIdentifier
+        if (tableIdentifier != null && requiredTrait == UpdateKindTrait.ONLY_UPDATE_AFTER) {
+          val catalogName = tableIdentifier.getCatalogName
+          val catalog = context.getCatalogManager.getCatalog(catalogName).orElse(null)
+          val catalogTable = normalize.table
+          if (ManagedTableListener.isManagedTable(catalog, catalogTable)) {
+            // if requiredTrait is ONLY_UPDATE_AFTER and table is ManagedTable,
+            // we can eliminate current normalize stage,
+            // cuz ManagedTable has preserved complete delete messages.
+            val input = if(normalize.getInput.isInstanceOf[StreamPhysicalExchange]) {
+              normalize.getInput.asInstanceOf[StreamPhysicalExchange].getInput
```

Comment:
Removing `StreamPhysicalExchange` is something dangerous, If there are two exchanges together, any problems?